src/rpc/connection/crypto.c

```diff
@@ -0,0 +1,323 @@
+#define CRYPTO_PREFIX_SPLONEBOXSERVER "splonebox-server"
+
+#define CRYPTO_ID_TUNNEL_SERVER "rZQTd2nT"
+#define CRYPTO_ID_TUNNEL_CLIENT "oqQN2kaT"
+#define CRYPTO_ID_MESSAGE_SERVER "rZQTd2nM"
+#define CRYPTO_ID_MESSAGE_CLIENT "oqQN2kaM"
+
+static unsigned char serverlongtermsk[32];
+
+int crypto_init(void)
+{
+  if (filesystem_load(".keys/server-long-term", serverlongtermsk,
+      sizeof serverlongtermsk) == -1) {
+    return (-1);
+  }
+
+  return (0);
+}
+
+static void nonce_update(struct crypto_context *cc)
+{
+  cc->nonce += 2;
+  if (cc->nonce)
+    return;
+  /* very unlikely */
+  LOG_ERROR("nonce space expired");
+  exit(1);
+}
+
+int byte_isequal(const void *yv, long long ylen, const void *xv)
+{
+  const unsigned char *y = yv;
+  const unsigned char *x = xv;
+  unsigned char diff = 0;
+
+  while (ylen > 0) {
+    diff |= (*y++ ^ *x++);
+    --ylen;
+  }
+
+  return (256 - (unsigned int) diff) >> 8;
+}
+
+void uint64_pack(unsigned char *y, uint64_t x)
+{
+  *y++ = (unsigned char) x; x >>= 8;
+  *y++ = (unsigned char) x; x >>= 8;
+  *y++ = (unsigned char) x; x >>= 8;
+  *y++ = (unsigned char) x; x >>= 8;
+  *y++ = (unsigned char) x; x >>= 8;
+  *y++ = (unsigned char) x; x >>= 8;
+  *y++ = (unsigned char) x; x >>= 8;
+  *y++ = (unsigned char) x; x >>= 8;
+}
+
+uint64_t uint64_unpack(const unsigned char *x)
+{
+  uint64_t result;
+
+  result = x[7];
+  result <<= 8; result |= x[6];
+  result <<= 8; result |= x[5];
+  result <<= 8; result |= x[4];
+  result <<= 8; result |= x[3];
+  result <<= 8; result |= x[2];
+  result <<= 8; result |= x[1];
+  result <<= 8; result |= x[0];
+
+  return result;
+}
+
+int crypto_verify_header(unsigned char *data, uint64_t *length)
+{
+  if (!(byte_isequal(data, 8, CRYPTO_ID_MESSAGE_CLIENT)))
+    return (-1);
+
+  *length = uint64_unpack(data + 8);
+
+  if (*length < 40)
+    return (-1);
+
+  return (0);
+}
+
+int crypto_tunnel(struct crypto_context *cc, unsigned char *data,
+    outputstream *out)
+{
+  unsigned char servershorttermpk[32];
+  unsigned char servershorttermsk[32];
+  unsigned char clientshorttermpk[32];
+  unsigned char clientshortserverlong[32];
+  unsigned char nonce[crypto_box_NONCEBYTES];
+  unsigned char packet[72];
+  unsigned char allzeroboxed[96] = {0};
+  unsigned char servershorttermpk0padded[64] = {0};
+  unsigned char servershorttermpkboxed[64];
+  uint64_t packetnonce;
+  uint64_t length;
+
+  /* check if first 8 byte of packet identifier are correct */
+  if (!(byte_isequal(data, 8, CRYPTO_ID_TUNNEL_CLIENT)))
+    return (-1);
+
+  /* unpack nonce and check it's validity */
+  packetnonce = uint64_unpack(data + 16);
+
+  if (packetnonce <= cc->receivednonce || !ISODD(packetnonce))
+    return (-1);
+
+  /* init clientshorttermpk */
+  memcpy(clientshorttermpk, data + 24, 32);
+  crypto_box_beforenm(clientshortserverlong, clientshorttermpk,
+      serverlongtermsk);
+
+  /* read length */
+  length = uint64_unpack(data + 8);
+  if (length != 136)
+    return (-1);
+
```

Comment:
A verification of the 64 byte zeros is missing. If the padding is missing or invalid, the server must not respond.